File: Baekjoon/Python3/2178. 미로 탐색/solution.py
# https://www.acmicpc.net/problem/2178

from collections import deque
import sys

input = sys.stdin.readline

n, m = map(int, input().split())  # 세로, 가로 길이
maze = [list(map(int, input().rstrip())) for _ in range(n)]

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

# 미로의 최단거리문제는 dfs대신 bfs를 써야함
def bfs():  # 시작위치가 고정이기때문에 매개변수가 필요없음
    queue = deque()  # deque를 써서 큐를 구현함
    queue.append([0, 0])
    while queue:
        now_y, now_x = queue.popleft()
        for idx in range(4):
            ny, nx = now_y + dy[idx], now_x + dx[idx]
            if ny < 0 or ny >= n or nx < 0 or nx >= m or maze[ny][nx] != 1:
                continue
            maze[ny][nx] = maze[now_y][now_x] + 1  # 큐에 넣기전에 방문처리를 해줘야 시간초과가 안남
            queue.append([ny, nx])


bfs()
print(maze[n - 1][m - 1])

File: Baekjoon/C++/2178. 미로 탐색/solution.cpp
// https://www.acmicpc.net/problem/2178

#include <iostream>
#include <vector>
#include <string>
#include <queue>

// 4방향으로 움직이는 배열
const int dy[4] = {0, 0, 1, -1};
const int dx[4] = {1, -1, 0, 0};

int main(int argc, char const *argv[])
{
    int n, m;
    std::cin >> n >> m;
    std::vector<std::vector<int>> distance(n, std::vector<int>(m, 0));
    std::vector<std::vector<char>> maze(n);
    std::string input;
    for (int i = 0; i < n; i++)
    {
        std::cin >> input;
        for (auto c : input)
        {
            maze[i].push_back(c);
        }
    }
    // BFS
    std::queue<std::pair<int, int>> q;
    // 시작지점 설정
    q.emplace(0, 0);
    distance[0][0] = 1;
    std::pair<int, int> cur;
    int x, y, newX, newY;
    while (!q.empty())
    {
        cur = q.front();
        x = cur.first;
        y = cur.second;
        q.pop();
        for (int idx = 0; idx < 4; idx++)
        {
            newX = x + dx[idx];
            newY = y + dy[idx];
            // 좌표가 미로를 벗어나지않고 벽도 아니고 한번도 방문하지 않은 좌표라면 큐에 좌표를 넣음
            if (0 <= newX && newX < m && 0 <= newY && newY < n && maze[newY][newX] == '1' && distance[newY][newX] == 0)
            {
                q.emplace(newX, newY);
                distance[newY][newX] = distance[y][x] + 1;
            }
        }
    }

    std::cout << distance[n - 1][m - 1];
    return 0;
}
